src/phongo_apm.c

```diff
@@ -262,6 +262,108 @@ static void phongo_apm_command_failed(const mongoc_apm_command_failed_t* event)
 	FREE_HASHTABLE(subscribers);
 }
 
+static void phongo_apm_server_changed(const mongoc_apm_server_changed_t* event)
+{
+	mongoc_client_t*                 client;
+	HashTable*                       subscribers;
+	php_phongo_serverchangedevent_t* p_event;
+	zval                             z_event;
+	const mongoc_host_list_t*        host_list;
+
+	client      = mongoc_apm_server_changed_get_context(event);
+	subscribers = phongo_apm_get_subscribers_to_notify(php_phongo_sdamsubscriber_ce, client);
+
+	/* Return early if there are no APM subscribers to notify */
+	if (zend_hash_num_elements(subscribers) == 0) {
+		goto cleanup;
+	}
+
+	object_init_ex(&z_event, php_phongo_serverchangedevent_ce);
+	p_event = Z_SERVERCHANGEDEVENT_OBJ_P(&z_event);
+
+	host_list = mongoc_apm_server_changed_get_host(event);
+	memcpy(&p_event->host, &host_list->host, BSON_HOST_NAME_MAX + 1);
+	p_event->port = host_list->port;
+
+	mongoc_apm_server_changed_get_topology_id(event, &p_event->topology_id);
+	p_event->new_server_description = mongoc_server_description_new_copy(mongoc_apm_server_changed_get_new_description(event));
+	p_event->old_server_description = mongoc_server_description_new_copy(mongoc_apm_server_changed_get_previous_description(event));
+
+	phongo_apm_dispatch_event(subscribers, "serverChanged", &z_event);
+	zval_ptr_dtor(&z_event);
+
+cleanup:
+	zend_hash_destroy(subscribers);
+	FREE_HASHTABLE(subscribers);
+}
+
+static void phongo_apm_server_closed(const mongoc_apm_server_closed_t* event)
+{
+	mongoc_client_t*                client;
+	HashTable*                      subscribers;
+	php_phongo_serverclosedevent_t* p_event;
+	zval                            z_event;
+	const mongoc_host_list_t*       host_list;
+
+	client      = mongoc_apm_server_closed_get_context(event);
+	subscribers = phongo_apm_get_subscribers_to_notify(php_phongo_sdamsubscriber_ce, client);
+
+	/* Return early if there are no APM subscribers to notify */
+	if (zend_hash_num_elements(subscribers) == 0) {
+		goto cleanup;
+	}
+
+	object_init_ex(&z_event, php_phongo_serverclosedevent_ce);
+	p_event = Z_SERVERCLOSEDEVENT_OBJ_P(&z_event);
+
+	host_list = mongoc_apm_server_closed_get_host(event);
+	memset(p_event->host, 0, sizeof(p_event->host));
+	strncpy(p_event->host, host_list->host, sizeof(p_event->host) - 1);
```

Comment:
I'm getting a build error on the GitHub CI using `strncpy` (https://github.com/mongodb/mongo-php-driver/pull/1251/checks?check_run_id=3373304261):

> In function ‘strncpy’,
    inlined from ‘phongo_apm_server_closed’ at /home/runner/work/mongo-php-driver/mongo-php-driver/src/phongo_apm.c:321:2:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:106:10: error: ‘__builtin_strncpy’ output may be truncated copying 255 bytes from a string of length 255 [-Werror=stringop-truncation]

Reply:
I believe https://stackoverflow.com/a/50198710 discusses the exact situation. I suggest reading through that question and the selected answer to understand what's going on here.

That said, we can also use [`bson_strncpy`](http://mongoc.org/libbson/current/bson_strncpy.html) here instead. That appears to be what libmongoc uses internally to copy the same field (see `_mongoc_host_list_push`). And `bson_strncpy` has the added benefit of always assigning a null byte to the end of the buffer (something `strncpy` does not do on its own).